src/serialize_utils.cpp

```diff
@@ -70,14 +72,24 @@ namespace sparrow_ipc
         return utils::align_to_8(total_size);
     }
 
-    std::size_t calculate_record_batch_message_size(const sparrow::record_batch& record_batch)
+    std::size_t calculate_record_batch_message_size(const sparrow::record_batch& record_batch, std::optional<org::apache::arrow::flatbuf::CompressionType> compression)
     {
         // Build the record batch message to get its exact metadata size
-        flatbuffers::FlatBufferBuilder record_batch_builder = get_record_batch_message_builder(record_batch);
+        flatbuffers::FlatBufferBuilder record_batch_builder = get_record_batch_message_builder(record_batch, compression);
         const flatbuffers::uoffset_t record_batch_len = record_batch_builder.GetSize();
 
-        // Calculate body size (already includes 8-byte alignment for each buffer)
-        const int64_t body_size = calculate_body_size(record_batch);
+        std::size_t actual_body_size = 0;
+        if (compression.has_value())
+        {
+            // If compressed, the body size is the sum of compressed buffer sizes + original size prefixes + padding
+            auto [compressed_body, compressed_buffers] = generate_compressed_body_and_buffers(record_batch, compression.value());
+            actual_body_size = compressed_body.size();
+        }
+        else
+        {
+            // If not compressed, the body size is the sum of uncompressed buffer sizes with padding
+            actual_body_size = static_cast<std::size_t>(calculate_body_size(record_batch));
+        }
```

Comment:
The same method should be able to handle both the compressed and uncompressed case, _especially_ as you want to transmit non-compressible data uncompressed anyway.